src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1023,12 +1023,22 @@ public static List<TSource> ToList<TSource>(this IEnumerable<TSource> source)
 
         public static Dictionary<TKey, TSource> ToDictionary<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector)
         {
-            return ToDictionary<TSource, TKey, TSource>(source, keySelector, IdentityFunction<TSource>.Instance, null);
+            return ToDictionary<TSource, TKey>(source, keySelector, null);
         }
 
         public static Dictionary<TKey, TSource> ToDictionary<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey> comparer)
         {
-            return ToDictionary<TSource, TKey, TSource>(source, keySelector, IdentityFunction<TSource>.Instance, comparer);
+            if (source == null) throw Error.ArgumentNull("source");
+            if (keySelector == null) throw Error.ArgumentNull("keySelector");
+
+            int capacity = 0;
+
+            if (source is TSource[]) capacity = (source as TSource[]).Length; // Cast to Array is much faster than cast to ICollection
+            else if (source is ICollection<TSource>) capacity = (source as ICollection<TSource>).Count;
```

Comment:
@stephentoub, sorry for the poor wording. I meant that the removal of the cast to Array would give a small gain (when source collection is not an array) compared with the code comprising that cast _from the current PR_.

@JonHanna, I see my mistake. I did not test cast to array with a wide range of types.

I've updated performance test to be closer to the code from this PR. Now it estimates a relative loss from adding a cast to array before cast to ```ICollection<T>```: https://gist.github.com/ikopylov/f3e6e1d210c848e06c2a

Source | (time of cast to ```ICollection<T>``` / time of cast to ```T[]``` then to ```ICollection<T>```) * 100.0 (%)
--------------|-------------------------
Enumerable.Repeat((object)null, 3)|72,53
Enumerable.Range(0, 2).Select(i => (object)i)|74,33
new object[0].Where(i => false)|69,85
new object[0]|426,53
new object[20]|427,28
new List<object>()|73,77
new HashSet<object>()|70,00
new List<string>()|78,19
new HashSet<string>()|82,52
new string[0]|137,21
new string[20]|136,85
new List<object>().AsReadOnly()|79,75
new List<string>().AsReadOnly()|77,71

So the loss is about 25%. 

More important question here is: how often in practice ```ToDictionary``` used with an array as source collection? I think this is a rare case. So I should agree with @stephentoub and eliminate the additional cast to array.


Note: @JonHanna, your code of the test from gist is not correct. The cast will be moved out from the loop during the optimization. But somehow your results is close to my with the fixed code.